File: src/app/beacon_service.js
const { createUuidV7 } = require('./uuid_v7');

class BeaconService {
  constructor({ idGenerator = createUuidV7 } = {}) {
    this.idGenerator = idGenerator;
    this.monitoring = false;
    this.facilityId = null;
    this.lastKnownMinor = null;
  }

  startMonitoring(authState) {
    if (!authState || !authState.canStartMonitoring()) {
      throw new Error('facility_id must be resolved before monitoring starts');
    }

    this.monitoring = true;
    this.facilityId = authState.facilityId;
  }

  static pickBestReading(rangingReadings = []) {
    const candidates = rangingReadings
      .filter((reading) => Number.isInteger(reading.minor))
      .filter((reading) => typeof reading.rssi === 'number' && reading.rssi > -90)
      .sort((left, right) => right.rssi - left.rssi);

    return candidates[0] ?? null;
  }

  static pickMinor(rangingReadings = []) {
    return BeaconService.pickBestReading(rangingReadings)?.minor ?? null;
  }

  _assertMonitoring() {
    if (!this.monitoring) {
      throw new Error('monitoring is not active');
    }
  }

  _createMovement({ staffId, action, minor, timestamp, rssi = null }) {
    return {
      id: this.idGenerator(),
      staffId,
      minor,
      action,
      timestamp,
      rssi
    };
  }

  handleEnterRegion({ staffId, timestamp, rangingReadings = [] }) {
    this._assertMonitoring();

    const bestReading = BeaconService.pickBestReading(rangingReadings);
    const minor = bestReading?.minor ?? this.lastKnownMinor;
    const rssi = bestReading?.rssi ?? null;

    if (Number.isInteger(minor)) {
      this.lastKnownMinor = minor;
    }

    return this._createMovement({
      staffId,
      action: 'ENTER',
      minor,
      timestamp,
      rssi
    });
  }

  handleExitRegion({ staffId, timestamp }) {
    this._assertMonitoring();

    return this._createMovement({
      staffId,
      action: 'EXIT',
      minor: this.lastKnownMinor,
      timestamp,
      rssi: null
    });
  }
}

module.exports = {
  BeaconService
};

File: lib/app/domain/beacon_service.dart
import 'auth_state.dart';
import 'beacon_reading.dart';
import 'movement_log.dart';
import 'uuid_v7.dart';

typedef BeaconIdGenerator = String Function();

class BeaconService {
  BeaconService({BeaconIdGenerator? idGenerator})
      : _idGenerator = idGenerator ?? createUuidV7;

  final BeaconIdGenerator _idGenerator;

  bool _isMonitoring = false;
  int? _facilityId;
  int? _lastKnownMinor;

  bool get isMonitoring => _isMonitoring;
  int? get facilityId => _facilityId;
  int? get lastKnownMinor => _lastKnownMinor;

  void startMonitoring(AuthState authState) {
    if (!authState.canStartMonitoring()) {
      throw StateError('facility_id must be resolved before monitoring starts');
    }
    _isMonitoring = true;
    _facilityId = authState.facilityId;
  }

  void stopMonitoring() {
    _isMonitoring = false;
    _lastKnownMinor = null;
  }

  void updateLastKnownMinor(int? minor) {
    if (minor != null) {
      _lastKnownMinor = minor;
    }
  }

  static BeaconReading? pickBestReading(List<BeaconReading> rangingReadings) {
    final candidates =
        rangingReadings.where((reading) => reading.rssi > -90).toList();
    if (candidates.isEmpty) {
      return null;
    }
    candidates.sort((left, right) => right.rssi.compareTo(left.rssi));
    return candidates.first;
  }

  static int? pickMinor(List<BeaconReading> rangingReadings) {
    return pickBestReading(rangingReadings)?.minor;
  }

  MovementLog handleEnterRegion({
    required String staffId,
    required DateTime timestamp,
    List<BeaconReading> rangingReadings = const [],
  }) {
    _assertMonitoring();
    final bestReading = pickBestReading(rangingReadings);
    final minor = bestReading?.minor ?? _lastKnownMinor;

    if (minor != null) {
      _lastKnownMinor = minor;
    }

    return MovementLog(
      id: _idGenerator(),
      staffId: staffId,
      minor: minor,
      action: 'ENTER',
      timestamp: timestamp,
      rssi: bestReading?.rssi,
    );
  }

  MovementLog handleExitRegion({
    required String staffId,
    required DateTime timestamp,
  }) {
    _assertMonitoring();
    return MovementLog(
      id: _idGenerator(),
      staffId: staffId,
      minor: _lastKnownMinor,
      action: 'EXIT',
      timestamp: timestamp,
      rssi: null,
    );
  }

  void _assertMonitoring() {
    if (!_isMonitoring) {
      throw StateError('monitoring is not active');
    }
  }
}
